src/Chainweb/Mempool/InMemTypes.hs

```diff
@@ -47,8 +47,10 @@ data InMemConfig t = InMemConfig {
     _inmemTxCfg :: {-# UNPACK #-} !(TransactionConfig t)
   , _inmemTxBlockSizeLimit :: !GasLimit
   , _inmemMaxRecentItems :: {-# UNPACK #-} !Int
-    -- Here Nothing means 'OK to insert'
-  , _inmemPreInsertCheck :: !(V.Vector t -> IO (V.Vector (Maybe InsertError)))
+  , _inmemPreInsertPureChecks :: t -> Either InsertError t
+  , _inmemPreInsertBatchChecks
+        :: V.Vector (TransactionHash, t)
+        -> IO (V.Vector (Either (TransactionHash, InsertError) (TransactionHash, t)))
```

Comment:
One-off and batch checks are now split. This increases the chance of being able to short-circuit.